### ISSUE TITLE
Zero Output from trained model

### ISSUE THREAD

**Issue by QasimWani** [pytorch/captum#617]
Every time I run an attribution function such as DeepLIFT or Integrated Gradients on a pre-trained model, I get a tensor of 0 values. This is different from a randomly initialized model which gives me non-zero attributions. 
My model is quite trivial as shown below:
```python

class SeedNetwork(nn.Module):
    """3 layer ANN for prediction of minimum number among C numbers"""
    def __init__(self, C=5, fc1=16, fc2=8):
        """ C represents number of classes """
        super(SeedNetwork, self).__init__()
        self.fc1 = nn.Linear(C, fc1)
        self.dp1 = nn.Dropout(p=0.3)
        self.fc2 = nn.Linear(fc1, fc2)
        self.dp2 = nn.Dropout(p=0.3)
        self.output = nn.Linear(fc2, C)
    
    def forward(self, x):
        """ Forward pass with input x of shape N x C """
        #Layer 1
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dp1(x)
        #Layer 2
        x = self.fc2(x)
        x = F.relu(x)
        x = self.dp2(x)
        #Output Layer
        x = torch.sigmoid(self.output(x))
        return x
```  

Running the Integrated Gradient using the documentation as shown below:
```python
# defining baselines for each input tensor
baseline = torch.zeros(10, 5)

# defining and laoding existing model
model = SeedNetwork()
model.load_state_dict(torch.load("../model/seedNetwork.pth"), strict=True)
model.eval()

# defining sample input
input = torch.rand(10, 5)

# defining and applying integrated gradients on SeedModel and the
ig = IntegratedGradients(model)
attributions, approximation_error = ig.attribute(inputs=input,
                                                 baselines=baseline,
                                                 return_convergence_delta=True,
                                                 target=0,
                                                 n_steps=50)
```
The output on this trained model is a tensor of 0 values as mentioned above:

```python
tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]], dtype=torch.float64)
```

#### What could this imply?
The overall model test accuracy is about ~90%. So, not entirely sure why I'm getting these results?
Would appreciate any help.
Thanks,
Qasim

**Comment by bilalsal** [comment]
Hi @QasimWani,

is it possible that the weights at one of the FC layers are zero?
Maybe just define [forward hooks](https://blog.paperspace.com/pytorch-hooks-gradient-clipping-debugging/)  and check the output of these layers for different inputs.

Another way to have a clue is to compute the attribution using one of the simpler algorithms(e.g. Input X Gradient, Guided Backprop, and Deconvolution, check [here](https://captum.ai/api/)).

Hope this helps

**Comment by QasimWani** [comment]
hey @bilalsal 
thanks for the reply. turns out the error was not from the model having zero weights but because the random input i was passing was seen as an error of sorts due to overfitting of model. when i pass in input from test/train set of integers, everything works fine. but when i use torch.randn of numbers 0-1, it all becomes zero. this is not true of the prediction itself which give correct results, but the attributions change for fractions as features as compared with integers as feature inputs.
any idea on what may be causing this subtle issue?

**Comment by bilalsal** [comment]
Hi @QasimWani ,

it is possible that the ReLU layers in your model are zeroing the output of the FC layers. If the bias terms in your 1st FC layer happens to be negative, then the output for this layer will likely be negative for a random input since the channels could probably cancel out each other.
Replacing ReLU by sigmoid or leaky ReLU should mitigate this issue, though unsure if this would fit what your model intends to accomplish.

Hope this helps

**Comment by QasimWani** [comment]
fixed it, thanks!

**Comment by tommyfuu** [comment]
hey! was wondering how you fixed this? thanks!
